### PR TITLE
Make sequential reductions use same pattern as OpenMP

### PR DESCRIPTION
**Pull request by davidbeckingsale**
This removes the need for the `m_blockdata` array, and lets us perform reductions on any type `T` that supports the appropriate reduction operator.

I'm only supporting sequential reductions in this PR and will make the changes for Cilk and CUDA separately.

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
@trws did you have any thoughts on mapping this pattern to Cilk and CUDA?

File: include/RAJA/exec-sequential/reduce_sequential.hxx [hunk 1]
```diff
@@ -146,12 +144,8 @@ private:
...
   ReduceMin<seq_reduce, T>();
 
-  bool m_is_copy;
-  int m_myID;
-
-  T m_reduced_val;
-
-  CPUReductionBlockDataType* m_blockdata;
+  const my_type * parent;
+  mutable T val;
```
Comment (rhornung67) [inline include/RAJA/exec-sequential/reduce_sequential.hxx]: Can we continue to use the convention that class member variables have "m_" prefix?

File: include/RAJA/exec-sequential/reduce_sequential.hxx [hunk 2]
```diff
@@ -79,30 +79,24 @@ namespace RAJA
 template <typename T>
 class ReduceMin<seq_reduce, T>
 {
+  using my_type = ReduceMin<seq_reduce, T>;
```
Comment (rhornung67) [inline include/RAJA/exec-sequential/reduce_sequential.hxx]: If you define this type alias here, should you use it everywhere you can; e.g., copy-ctor arg, etc.?

File: include/RAJA/exec-sequential/reduce_sequential.hxx [hunk 3]
```diff
@@ -111,8 +105,8 @@ public:
   //
   ~ReduceMin<seq_reduce, T>()
   {
-    if (!m_is_copy) {
-      releaseCPUReductionId(m_myID);
+    if (parent) {
+      parent->min(val);
```
Comment (rhornung67) [inline include/RAJA/exec-sequential/reduce_sequential.hxx]: How does this avoid a potential race condition?

Comment (davidbeckingsale) [inline include/RAJA/exec-sequential/reduce_sequential.hxx]: It's a sequential reduction, so no race conditions?

File: include/RAJA/exec-sequential/reduce_sequential.hxx (hunk 1 above)
Comment (davidbeckingsale) [inline include/RAJA/exec-sequential/reduce_sequential.hxx]: Sure, I can fix that up. I will do that in the OpenMP reductions also.

**Comment by trws** [comment]
For cilk, it would basically just be using their hyperobjects to get copy construction on steal rather than explicitly copy-constructing on thread creation.  Just wrapping around their reducers actually does the same thing though, so if we've already got that I'm not sure it's worth the extra effort.

For CUDA, we had at least one working prototype that did this same thing in place of using dynamically allocated shared memory.  It still used shared memory in the destructor, but only there, so it could use statically-sized local shared memory instead of dynamically allocating it.  IIRC, when the current CUDA reductions were going in, both versions were tested and the current version was found to be slightly slower by itself but produce less register pressure and so better overall.  I'm not sure what happened to the prototypes, anyone know if they're still around somewhere? I'm sure I could dig up the basic sum one I threw together but I'm pretty sure someone did a more complete job of it to test it.

File: include/RAJA/exec-sequential/reduce_sequential.hxx (hunk 3 above)
Comment (rhornung67) [inline include/RAJA/exec-sequential/reduce_sequential.hxx]: Doh! I should pay closer attention to what I'm looking at...

Comment (davidbeckingsale) [inline include/RAJA/exec-sequential/reduce_sequential.hxx]: 😄

**Comment by jonesholger** [comment]
@trws With respect to the CUDA variant, @MrBurmark weaved it into a prototype he was testing, and had at one point in mind a  policy switch variant if you could tolerate the increase in register pressure. He had some findings on this, which if they were benign, we would have completely switched over to this, since I believe this pattern was actually more semantically correct (or supported more use cases) in addition to being slightly faster.

**Comment by jonesholger** [comment]
@davidbeckingsale just verifying -- were your constraints on adapting CILK and CUDA part of this PR?

**Comment by davidbeckingsale** [comment]
@jonesholger it would be good to have the cilk reductions done too in this PR, but I don't mind waiting for the CUDA ones.

**Comment by davidbeckingsale** [comment]
Status update on this PR - I've decided to punt on the cilk reductions for now and just get the sequential stuff merged.

I've addressed @rhornung67's comments re: member variable naming in both the OpenMP and sequential versions.

**Pull request merged**